### PR TITLE
build: Check usages of #if defined(...)

### PR DESCRIPTION
**Pull request by brokenprogrammer**
I tried to go over #16419.

Based on #16344 I interpreted that the AC_DEFINE like the following: 
```
AC_DEFINE(USE_EXTERNAL_ASM, 1, [Define this symbol if an external (non-inline) assembly implementation is used])
```
Results in value defines which also #16547 suggests.

If I've missunderstood anything or if there is more `#if defined` or `#ifdef` that should be `#if` then please let me know and I will update the pull request.

### CODE REVIEW

**Comment by brokenprogrammer** [comment]
The linting failed because of `FAIL: subtree directory was touched without subtree merge`. I guess this is because I touched some external library code. Should I just revert the changes that affected secp256k1?

**Comment by laanwj** [comment]
> The linting failed because of FAIL: subtree directory was touched without subtree merge. I guess this is because I touched some external library code. Should I just revert the changes that affected secp256k1?

Yes, it's not possible to make changes to subtrees through a PR in this repository, it needs to go through the appropriate upstream.

**Comment by brokenprogrammer** [comment]
> > The linting failed because of FAIL: subtree directory was touched without subtree merge. I guess this is because I touched some external library code. Should I just revert the changes that affected secp256k1?
> 
> Yes, it's not possible to make changes to subtrees through a PR in this repository, it needs to go through the appropriate upstream.

I see, thank you for the information. I have reverted the files for secp256k1.

**Comment by fanquake** [comment]
> I have reverted the files for secp256k1.

You need to remove the changes entirely, no reversion commits. You'll also need to write a proper / useful commit message for the actual change. I'd suggest reading through https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md.

**Comment by brokenprogrammer** [comment]
> > I have reverted the files for secp256k1.
> 
> You need to remove the changes entirely, no reversion commits. You'll also need to write a proper / useful commit message for the actual change. I'd suggest reading through https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md.
> 

Sorry about that, I've published a new change which I hope corrects my mistakes.

**Comment by luke-jr** [comment]
AC_DEFINE only defines if it's actually called. Conditional calls to AC_DEFINE (such as those setting it to a constant `1`, or not setting a value at all) need `defined()`/`#ifdef` AIUI

Looking through our defines, I don't see anything that requires changes.

**Comment by luke-jr** [comment]
Correction: LevelDB is using `#if defined(HAVE_O_CLOEXEC)`, which we are defining to `0` on systems without it - this leads to a build failure, and should be fixed.

**Comment by brokenprogrammer** [comment]
> AC_DEFINE only defines if it's actually called. Conditional calls to AC_DEFINE (such as those setting it to a constant `1`, or not setting a value at all) need `defined()`/`#ifdef` AIUI
> 
> ~Looking through our defines, I don't see anything that requires changes.~ Edit: was looking at an old commit

With that said I think I've missunderstood the task? Should I remove all my previous changes and perhaps only address what you mentioned with `HAVE_O_CLOEXEC`?

**Comment by fanquake** [comment]
> Should I remove all my previous changes and perhaps only address what you mentioned with HAVE_O_CLOEXEC?

No, that issue has now been resolved in master. However you can rebase / squash your changes here.

**Comment by fanquake** [comment]
@brokenprogrammer interested in following?

**Comment by brokenprogrammer** [comment]
@fanquake So the changes are fine and what is left is simply to merge latest master and squash the changes?

**Comment by fanquake** [comment]
@brokenprogrammer sorry for not following up. Changes should always be squashed. I will get to reviewing this shortly.

File: src/bench/verify_script.cpp [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 #include <bench/bench.h>
 #include <key.h>
-#if defined(HAVE_CONSENSUS_LIB)
+#if HAVE_CONSENSUS_LIB
```
Comment (hebasto) [inline src/bench/verify_script.cpp:7]: > Conditional calls to AC_DEFINE (such as those setting it to a constant `1`, or not setting a value at all) need `defined()`/`#ifdef` AIUI

Agree with @luke-jr.

In this particular case, it is conditional: https://github.com/bitcoin/bitcoin/blob/8126551d54ffd290ed5767248be4b3d19243787b/configure.ac#L1707-L1710

therefore, the current code is fine.

**Comment by brokenprogrammer** [comment]
@fanquake No worries, I've squashed and updated the commit message according to the contributing guidelines.

**Comment by hebasto** [comment]
@brokenprogrammer 
> I've squashed and updated the commit message according to the contributing guidelines.

A https://github.com/bitcoin/bitcoin/pull/25302#discussion_r1104616746 still needs to be addressed.

**Comment by hebasto** [comment]
Another bug found, which needs to be addressed.

The following code
https://github.com/bitcoin/bitcoin/blob/75f0e0b607cd7ff7afd56853eb34a2b285b22ad2/configure.ac#L1251-L1259
always defines the `HAVE_STRONG_GETAUXVAL` macro.

So the following `#if defined` directives can't handle it properly:
https://github.com/bitcoin/bitcoin/blob/75f0e0b607cd7ff7afd56853eb34a2b285b22ad2/src/randomenv.cpp#L56-L58 and https://github.com/bitcoin/bitcoin/blob/75f0e0b607cd7ff7afd56853eb34a2b285b22ad2/src/randomenv.cpp#L328-L348

File: src/crypto/sha256.cpp [hunk 2]
```diff
@@ -604,7 +604,7 @@ std::string SHA256AutoDetect()
         have_x86_shani = (ebx >> 29) & 1;
     }
 
-#if defined(ENABLE_X86_SHANI) && !defined(BUILD_BITCOIN_INTERNAL)
+#if ENABLE_X86_SHANI && !defined(BUILD_BITCOIN_INTERNAL)
```
Comment (TheCharlatan) [inline src/crypto/sha256.cpp:607]: I think this is only defined conditionally.

**Comment by hebasto** [comment]
I've re-reviewed this PR and none of the suggested changes seem required to me.

More details see in https://github.com/bitcoin/bitcoin/issues/16419#issuecomment-1552877492.

**Comment by fanquake** [comment]
Going to close for now. We can continue any discussion in #16419.

**Pull request closed without merging**